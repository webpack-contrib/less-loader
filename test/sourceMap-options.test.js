import path from "path";
import fs from "fs";

import {
  compile,
  getCodeFromBundle,
  getCodeFromLess,
  getCompiler,
  getErrors,
  getWarnings,
} from "./helpers";

describe('"sourceMap" options', () => {
  it('should generate source maps when value is "true"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(testId, {
      sourceMap: true,
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when the "devtool" value is "source-map"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(
      testId,
      {},
      {
        devtool: "source-map",
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when value is "true" and the "devtool" value is "false"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(
      testId,
      {
        sourceMap: true,
      },
      {
        devtool: false,
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when value has "false" value, but the "lessOptions.sourceMap.outputSourceFiles" is "true"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(testId, {
      sourceMap: false,
      lessOptions: {
        sourceMap: { outputSourceFiles: true },
      },
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when value is "false"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(testId, {
      sourceMap: false,
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when the "devtool" value is "false"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(
      testId,
      {},
      {
        devtool: false,
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when value is "false" and the "devtool" value is "source-map"', async () => {
    const testId = "./source-map.less";
    const compiler = getCompiler(
      testId,
      {
        sourceMap: false,
      },
      {
        devtool: "source-map",
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromLess = await getCodeFromLess(testId);
    const { css, map } = codeFromBundle;

    expect(css).toBe(codeFromLess.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
