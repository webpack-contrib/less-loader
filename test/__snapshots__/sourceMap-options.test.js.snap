// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": errors 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": source map 1`] = `
Object {
  "mappings": "AAAA;ACEA;EDDC,cAAA;;ACCD;EACE,YAAA",
  "names": Array [],
  "sourceRoot": "",
  "sources": Array [
    "test/fixtures/node_modules/some/module.less",
    "test/fixtures/source-map.less",
  ],
  "sourcesContent": Array [
    ".modules-dir-some-module {
	color: hotpink;
}
",
    "@import \\"some/module\\";

#it-works:extend(.modules-dir-some-module) {
  margin: 10px;
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": warnings 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value has "false" value, but the "lessOptions.sourceMap.outputSourceFiles" is "true": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when value has "false" value, but the "lessOptions.sourceMap.outputSourceFiles" is "true": errors 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value has "false" value, but the "lessOptions.sourceMap.outputSourceFiles" is "true": source map 1`] = `
Object {
  "mappings": "AAAA;ACEA;EDDC,cAAA;;ACCD;EACE,YAAA",
  "names": Array [],
  "sourceRoot": "",
  "sources": Array [
    "test/fixtures/node_modules/some/module.less",
    "test/fixtures/source-map.less",
  ],
  "sourcesContent": Array [
    ".modules-dir-some-module {
	color: hotpink;
}
",
    "@import \\"some/module\\";

#it-works:extend(.modules-dir-some-module) {
  margin: 10px;
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when value has "false" value, but the "lessOptions.sourceMap.outputSourceFiles" is "true": warnings 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": errors 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": source map 1`] = `
Object {
  "mappings": "AAAA;ACEA;EDDC,cAAA;;ACCD;EACE,YAAA",
  "names": Array [],
  "sourceRoot": "",
  "sources": Array [
    "test/fixtures/node_modules/some/module.less",
    "test/fixtures/source-map.less",
  ],
  "sourcesContent": Array [
    ".modules-dir-some-module {
	color: hotpink;
}
",
    "@import \\"some/module\\";

#it-works:extend(.modules-dir-some-module) {
  margin: 10px;
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": warnings 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value is "true": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when value is "true": errors 1`] = `Array []`;

exports[`"sourceMap" options should generate source maps when value is "true": source map 1`] = `
Object {
  "mappings": "AAAA;ACEA;EDDC,cAAA;;ACCD;EACE,YAAA",
  "names": Array [],
  "sourceRoot": "",
  "sources": Array [
    "test/fixtures/node_modules/some/module.less",
    "test/fixtures/source-map.less",
  ],
  "sourcesContent": Array [
    ".modules-dir-some-module {
	color: hotpink;
}
",
    "@import \\"some/module\\";

#it-works:extend(.modules-dir-some-module) {
  margin: 10px;
}
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when value is "true": warnings 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": errors 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": warnings 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": errors 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": warnings 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when value is "false": css 1`] = `
".modules-dir-some-module,
#it-works {
  color: hotpink;
}
#it-works {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when value is "false": errors 1`] = `Array []`;

exports[`"sourceMap" options should not generate source maps when value is "false": warnings 1`] = `Array []`;
